src/rsa/verification.rs

```diff
@@ -61,6 +60,86 @@ rsa_pkcs1!(RSA_PKCS1_3072_8192_SHA384, 3072, &super::RSA_PKCS1_SHA384,
            "Verification of signatures using RSA keys of 3072-8192 bits,
             PKCS#1.5 padding, and SHA-384.");
 
+extern {
+    fn GFp_rsa_check_modulus_and_exponent(n: *const bn::BIGNUM,
+                                          e: *const bn::BIGNUM,
+                                          min_bits: c::size_t,
+                                          max_bits: c::size_t) -> c::int;
+}
+
+fn rsa_public_decrypt(out: &mut [u8], public_key_n: &bn::PositiveInteger,
+                      public_key_e: &bn::PositiveInteger, ciphertext: &[u8],
+                      min_bits: usize, max_bits: usize)
+                      -> Result<(), error::Unspecified> {
+    // We define this closure as a translation of `goto`. We still want
```

Comment:
My main worry with starting with a literal transliteration of the C code into Rust is that we'd be stuck with needing to use the license that the C code uses. So, even though your approach here is very logical in showing that the Rust code is line-for-line the same the C code, that's actually undesirable. 

Of course, at the end of the day we need to do basically the same thing so there's going to be some similarity, but I'd rather get rid of this intermediate state completely, so that the licensing issue doesn't loom over us.

We know that we need to do the following things:
* Verify that the public modulus and exponent are valid.
* Check that the input and output lengths are exactly equal to the length in bytes of the public modulus.
* Verify that the input, when interpreted as a big-endian-encoded number, is less than the public modulus.
* Do the modular exponentiation.
* Encode the result of the exponentiation in big-endian.

For some of these steps, we already have Rust code available. For example we already have `PositiveInteger::from_be_bytes` so we should be able to already avoid using `GFp_BN_new()` and `GFp_BN_bin2bn()` directly. So, I suggest that you first write the Rust code for the function from scratch such that its body doesn't use any `unsafe` directly. If there is some function that we don't have Rust code for, like the modular exponentiation, then we can create a function, e.g. `fn mod_exp_vartime(in_out: &mut PositiveInteger, exponent: u64, modulus: &PositiveInteger) -> Result<(), error::Unspecified`, that uses `unsafe` under the hood to call one of the `GFp_` functions.

WDYT?